[3][Span logging] Span logging with SqlAlchemyStore - add status update logic

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/17070?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17070/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17070/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17070/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Add span logging with SqlAlchemyStore

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

dbczumar committed:
Add log_spans functionality to SqlAlchemyStore with concurrency safety

- Add spans table migration with Text columns for name/type/content
- Add computed duration_ns column for efficient duration filtering
- Add composite indexes for common query patterns (experiment_id, status, type)
- Implement log_spans and log_spans_async methods in SqlAlchemyStore
- Add race condition handling for concurrent trace creation using try-except
- Fix trace_state serialization bug in span.to_proto()
- Add comprehensive tests including concurrent trace creation test
- Update method signatures to accept experiment_id as first parameter

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor log_spans tests to use pytest.mark.parametrize

- Combine separate sync and async tests into parametrized tests
- Test both sync (log_spans) and async (log_spans_async) methods with same test logic
- Reduces code duplication and ensures consistent testing across both methods

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify test names in parametrized log_spans tests

- Remove unnecessary suffix for experiment names and trace IDs
- Each test run gets its own isolated database, so no naming conflicts
- Makes tests cleaner and easier to read

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Revert trace_state serialization and update exception types

- Revert trace_state.to_header() call in span.py's to_proto() method
- Update AbstractStore log_spans docstrings to use MlflowException instead of ValueError
- SqlAlchemyStore already raises MlflowException with INVALID_PARAMETER_VALUE error code

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update log_spans to update trace time range when new spans extend it

- Update trace timestamp_ms and execution_time_ms when new spans extend the time range
- This ensures the trace always reflects the full time span of all its spans
- Add test to verify trace time range is updated correctly when spans are added

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Make trace time range updates atomic to avoid race conditions

- Use SQLAlchemy's case expressions for atomic trace time updates
- This ensures concurrent span insertions correctly update trace time ranges
- Works across all supported databases without raw SQL
- Prevents race conditions where concurrent updates could lose time range extensions

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add comprehensive test for multiple log_spans calls updating trace times

- Test verifies trace start/end times are correctly updated across multiple calls
- Tests adding spans that extend the trace both earlier and later
- Tests that spans within the existing range don't change trace times
- Verifies all spans are saved correctly

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix trace_state handling and remove unnecessary hasattr checks

- Fix trace_state serialization in span.to_proto() to handle TraceState objects
- Remove defensive hasattr checks for _span and context - trust the API contract
- Handle trace_state extraction separately in log_spans to avoid JSON serialization issues
- Pop trace_state from span dict before JSON serialization since we handle it separately

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix log_spans trace_state handling for mocked tests

- Update log_spans to handle both TraceState objects and strings
- Fix test to mock trace_state as string (as expected by to_proto)
- Maintain compatibility with both production and test scenarios

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove trace_state column from spans table and related logic

- Remove trace_state column from database migration
- Remove trace_state field from SqlSpan model
- Remove trace_state handling from SqlAlchemyStore.log_spans
- Update tests to use empty TraceState objects
- Remove test_log_spans_multiple_calls_updates_trace_times test case

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add async parametrization to all log_spans tests and cleanup comments

- Add @pytest.mark.asyncio and @pytest.mark.parametrize('is_async', [False, True]) to all log_spans test functions
- Update all test functions to support both sync and async execution
- Add import for 'case' from sqlalchemy in test file
- Remove unnecessary inline comments from SqlAlchemyStore.log_spans implementation
- Fix pytest.raises block to satisfy linter requirements

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Set trace status from root span status when creating traces

When logging spans, if a root span (no parent) is found:
- Use root span's ERROR status for trace ERROR status
- Use root span's UNSET status for trace STATE_UNSPECIFIED status
- Use root span's OK status for trace OK status
- Also update trace status if it's currently STATE_UNSPECIFIED

This ensures trace status reflects the root span status rather than
always defaulting to OK.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update trace status logic for log_spans

- Default new traces to IN_PROGRESS status instead of OK
- Set trace status from root span when creating new traces
- Only update trace status from root span if current status is
  IN_PROGRESS or STATE_UNSPECIFIED (don't override finalized statuses)
- Map span statuses to trace statuses:
  - Span ERROR -> Trace ERROR
  - Span UNSET -> Trace STATE_UNSPECIFIED
  - Span OK -> Trace OK

This ensures traces start in a pending state and get updated based on
root span completion, while preserving finalized trace statuses.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor root span status extraction to helper method

Extract duplicated logic for determining trace status from root span
into a reusable _get_trace_status_from_root_span helper method.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix test expectation for trace status when span status is UNSET

When a span has UNSET status (the default OpenTelemetry status), the trace
status is correctly set to STATE_UNSPECIFIED rather than OK. Updated test
to match the actual behavior from _get_trace_status_from_root_span().

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix trace status mapping: UNSET span status should map to IN_PROGRESS

STATE_UNSPECIFIED should never be used for a trace that has logged spans.
When a span has UNSET status (the default OpenTelemetry status), the trace
should be IN_PROGRESS, not STATE_UNSPECIFIED. This makes much more sense
semantically - if we have spans but their status is unset, the trace is
still in progress.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Map UNSET span status to OK trace status

When root span has UNSET status (unexpected), assume the overall trace
status is OK rather than IN_PROGRESS. Updated tests to use realistic OK
status instead of UNSET for root spans, while still testing the UNSET
to OK mapping behavior.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Parametrize test_log_spans_sets_trace_status_from_root_span test

- Use pytest.mark.parametrize to test OK and ERROR status codes
- Keep separate test for UNSET status with clear indication it's unexpected
- Reduce code duplication in test implementation

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify trace status mapping logic in _get_trace_status_from_root_span

- Combine OK and UNSET cases since both return OK trace status
- Add clearer comment explaining the three valid span statuses (OK, ERROR, UNSET)
- Note that UNSET is unexpected in production but handled gracefully

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use constants for trace status values and add test helper functions

- Replace hardcoded string literals with TraceState and SpanStatusCode constants
- Add imports for SpanStatusCode and TraceState from mlflow.entities
- Create helper functions to deduplicate OTelReadableSpan creation in tests
- Refactor test functions to use the new helpers for cleaner code

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify span test helper functions for better usability

- Add create_test_span() helper that returns ready-to-use MLflow spans
- Use simpler, more intuitive parameter names (span_id, parent_id, start_ns)
- Significantly reduce boilerplate in tests
- Refactor verbose tests to use the new simpler helper
- Tests are now much more concise and easier to maintain

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy commented on tests/store/tracking/test_sqlalchemy_store.py:
```diff
@@ -6631,3 +6723,288 @@ def test_assessment_with_error(store_and_trace_info):
+
+    # Check trace status was updated to OK from root span
+    traces, _ = store.search_traces([experiment_id])
+    trace = [t for t in traces if t.request_id == trace_id][0]
```
```suggestion
    trace = next(t for t in traces if t.request_id == trace_id)
```

harupy replied:
the same comments on other `[t for t in traces if t.request_id == trace_id][0]`

harupy commented on tests/store/tracking/test_sqlalchemy_store.py:
```diff
@@ -6631,3 +6723,288 @@ def test_assessment_with_error(store_and_trace_info):
+):
+    """Test that trace status is updated from root span on subsequent logs."""
+    experiment_id = store.create_experiment("test_unspecified_update")
+    trace_id = "test_trace_unspec_" + str(uuid.uuid4().hex)
```
```suggestion
    trace_id = f"test_trace_unspec_{uuid.uuid4().hex}"
```

harupy commented on tests/store/tracking/test_sqlalchemy_store.py:
```diff
@@ -6631,3 +6723,288 @@ def test_assessment_with_error(store_and_trace_info):
+    else:
+        store.log_spans(experiment_id, [initial_span])
+
+    # Verify trace was created with OK status
+    traces, _ = store.search_traces([experiment_id])
+    trace = [t for t in traces if t.request_id == trace_id][0]
+    assert trace.state.value == "OK"
```
can we use `get_trace_info`?

harupy commented on tests/store/tracking/test_sqlalchemy_store.py:
```diff
@@ -131,6 +131,98 @@
 
 
+# Helper functions for span tests
+def create_mock_span_context(trace_id_num=12345, span_id_num=111):
```
can we add type hints?

harupy reviewed (approved):
LGTM once comments are addressed!

dbczumar committed:
Address PR review comments from @harupy

- Replace list comprehension with next() for finding traces
- Add return type hints to helper functions
- Use f-strings for trace_id formatting
- Use get_trace_info instead of search_traces where appropriate
- Add type hints to test function parameters

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use more specific SpanContext type hint instead of mock.Mock

The function returns a mock that behaves like a SpanContext, so using
trace_api.SpanContext as the return type hint is more descriptive than
the generic mock.Mock.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix trace_id length to respect database constraint

The trace_info table has a 50-character limit on the request_id column.
Shortened UUID hex strings from full 32 chars to 12 chars to ensure
test trace IDs stay under the limit.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use proper MLflow trace ID format in tests

Changed test trace IDs from custom prefixes to the standard MLflow
format: "tr-<32-char-hex>". This ensures tests use legitimate trace
IDs that match production format and are well within the 50-character
database constraint.

Signed-off-by: dbczumar <corey.zumar@databricks.com>
